### PR TITLE
Example of a datascience project with a couple of standard tools

### PR DESCRIPTION
**Pull request by laurensWe**
# Description

Standard data science project with the lifecycle blocks of such a project

Link all GitHub issues fixed by this PR.
Make sure to first open an issue, get community approval and only then create Pull Request to resolve it.
All Pull Requests must have an issue attached to them

Resolves #

## Checklist

Please check all the boxes that apply to this pull request using "x":

- [x]  I have tested the changes locally and verified that they work as expected.
- [x]  I have followed the project's coding conventions and style guidelines.
- [x]  I have rebased my branch onto the latest commit of the main branch.
- [x]  I have squashed or reorganized my commits into logical units.
- [x]  I have read, understood and agree to the [Developer Certificate of Origin](../blob/main/DCO.md), which this project utilizes.

### CODE REVIEW

File: example_datascience/__main__.py [hunk 1]
```diff
@@ -1,4 +1,220 @@
 import sys
+import pandas as pd
+import numpy as np
+from pycaret.classification import ClassificationExperiment, create_model, check_fairness
```
Comment (GuusjeJuijn) [inline example_datascience/__main__.py:4]: Pycaret only supports python 3.9, 3.10, 3.11, so that may become a problem here for Python 3.12 users? (I coudn't run the code because of that)

File: example_datascience/__main__.py [hunk 2]
```diff
@@ -1,4 +1,220 @@
...
+    cat_cols = [
+        "gender",
+        "nationality",
+        "sport",
+        "ind-debateclub",
+        "ind-programming_exp",
+        "ind-international_exp",
+        "ind-entrepeneur_exp",
+        "ind-exact_study",
+        "company",
```
Comment (GuusjeJuijn) [inline example_datascience/__main__.py:58]: The dataset contains information about 4 different companies, maybe it could be useful to also split the data into 4 subsets of these companies to find out whether there are different biases present across different companies

File: example_datascience/__main__.py [hunk 3]
```diff
@@ -1,4 +1,220 @@
...
+    ]
+
+    for col in cat_cols:
+        categories = df.groupby(col).size()
+        # print(categories)
+    # From the categorical features we see that the non-binary gender is underrepresented although it is from the true
+    # distribution of society at this moment. The same holds for nationality where Dutch is overrepresented. almost 3/4
+    # of the applicants got rejected which we need to take into account in our model as well.
+    return df
+
```
Comment (GuusjeJuijn) [inline example_datascience/__main__.py:69]: Would also be useful to add countplots in this EDA showing the decision (hired / not hired) compared to the sensitive attributes (in that way you can for example immediately see whether there is a gender bias in the data)

File: example_datascience/__main__.py [hunk 4]
```diff
@@ -1,4 +1,220 @@
...
+        "company",
+        "decision",
+    ]
+
+    for col in cat_cols:
+        categories = df.groupby(col).size()
+        # print(categories)
+    # From the categorical features we see that the non-binary gender is underrepresented although it is from the true
+    # distribution of society at this moment. The same holds for nationality where Dutch is overrepresented. almost 3/4
+    # of the applicants got rejected which we need to take into account in our model as well.
```
Comment (GuusjeJuijn) [inline example_datascience/__main__.py:67]: also add some information about the distribution of the target variable?

File: example_datascience/__main__.py [hunk 5]
```diff
@@ -1,4 +1,220 @@
...
+    return df
+
+
+def preprocessing_data(data):
```
Comment (GuusjeJuijn) [inline example_datascience/__main__.py:71]: Where do you split the data into a train and test set ? maybe good to do that before preprocessing the data

File: example_datascience/__main__.py [hunk 6]
```diff
@@ -1,4 +1,220 @@
...
+    :return:
+    """
+    return 0
+
+
+def convert_to_standard_dataset_for_aif360(df, target_label_name, scores_name=""):
+    protected_attributes = []
+
+    # columns from the dataset that we want to select for this Bias study
+    selected_features = ["gender", "age"]
```
Comment (GuusjeJuijn) [inline example_datascience/__main__.py:194]: maybe nationality could also be considered as a sensitive attribute?

File: example_datascience/__main__.py [hunk 7]
```diff
@@ -219,6 +436,104 @@ def main() -> int:
...
+    # # print(dp)
+    # # print(sr.by_group)
+    #
+    # # ## check fairness out of the box of pycaret
+    # # catboost_fairness = exp.check_fairness(catboost, sensitive_features=['gender'])
+    # # print(catboost_fairness)
+    #
+    # ## check mitigation of demographic disparity of gender via fairlearn
+    # # constraint = DemographicParity()
+    # # mitigator = ExponentiatedGradient(catboost, constraint)
```
Comment (GuusjeJuijn) [inline example_datascience/__main__.py:477]: Could also be interesting to try out a postprocessing bias mitigation algorithm instead of only using pre-processing bias mitigation

File: example_datascience/__main__.py [hunk 8]
```diff
@@ -1,4 +1,220 @@
...
+from sklearn.model_selection import cross_val_predict
+import joblib
+
+np.random.seed(69)
+
+
+def exploratory_data_analysis():
+    """
+    This function will do the exploratory data analysis (EDA) on the recruitment dataset of the University of Utrecht.
+    In most project this is done in a Jupyter Notebook as this makes interaction easier, but a slice of the EDA is still
```
Comment (GuusjeJuijn) [inline example_datascience/__main__.py:34]: projects*

**Review by GuusjeJuijn — COMMENTED** [review]
Looks good! I added some suggestions and minor details

File: example_datascience/__main__.py (hunk 1 above)
Comment (laurensWe) [reply example_datascience/__main__.py:4]: Yes that is true, hence i bumped the .python_version down to 3.11

File: example_datascience/__main__.py (hunk 2 above)
Comment (laurensWe) [reply example_datascience/__main__.py:58]: valid point!

Comment (laurensWe) [reply example_datascience/__main__.py:58]: As right now I would use it as feature, which doesn't make so much sense

File: example_datascience/__main__.py (hunk 3 above)
Comment (laurensWe) [reply example_datascience/__main__.py:69]: I reserved this for the evaluation of the model, but showing it here already gives us an idea that we should look later at the demographic disparity agree

File: example_datascience/__main__.py (hunk 7 above)
Comment (laurensWe) [reply example_datascience/__main__.py:477]: Yeah, I was thinking about this as well but I am doubting whether it would make this example project too large or not?

File: example_datascience/__main__.py (hunk 8 above)
Comment (GuusjeJuijn) [inline example_datascience/__main__.py:34]:
```suggestion
    In most projects this is done in a Jupyter Notebook as this makes interaction easier, but a slice of the EDA is still
```

File: example_datascience/__main__.py (hunk 5 above)
Comment (laurensWe) [reply example_datascience/__main__.py]: i will make this extra clear :)

**Review by ChristopherSpelt — COMMENTED** [review]
Looks good! Maybe add a small addition in the readme on how you implemented this, after the description that is already there in the readme. Running the main file does a lot of things, and also as you mentioned in your code comments some things are usually done in notebooks. So maybe it's good to add a small text explaining that while the normal workflow of data scientist is different, for convenience you made a single file with separate functions representing the steps you mentioned in the description.

File: README.md [hunk 9]
```diff
@@ -1,36 +1,40 @@
...
+3. evaluating models
+4. serving a result
+5. monitoring
 
-When creating a new Repository select this template repository as the base.
+Because building an algorithm is not linear and if unwanted behaviour is seen in any step of the progress it could be
+that you are returning to previous steps in the process. Because of this process, it is hard to define this in this
+project. Therefore, I have chosen comment out the chain-of-thought that one could have in order to f.e. mitigate bias
+or choosing a specific model. So when running `__main__.py` does not result in any output as it will make a model and
+does inference on it without anything to return.
```
Comment (ChristopherSpelt) [inline README.md:21]:
```suggestion
Building an algorithm is non linear, because of dependencies and cycles in the above mentioned steps. Because of this, it is hard to define this in this in a standard project template. Therefore, I have chosen to make a single Python file and comment out the chain-of-thought that one could have in order to e.g. mitigate bias
or choosing a specific model. So running `__main__.py` will make a model and
does inference on it without anything to return.
```

File: README.md [hunk 10]
```diff
@@ -1,36 +1,40 @@
...
-  * linux: `find . -type f -not -path "./.git/*" -exec  sed -i "s/python_project/new_name/g" {} \;`
-* rename the python_project/ folder to your project name
-* change author and name in pyproject.toml
-* change labels in Dockerfile to appropriate values
-* Verify the License used
-* Change publiccode.yml to your needs
+## Tools used
+
+Different frameworks are used for different stages in the project, for the exploratory data analyses `PyCaret` is used to
+very quickly check a bunch of models and in which direction to build the "production" model later. For the "production"
```
Comment (ChristopherSpelt) [inline README.md:31]:
```suggestion
Different frameworks are used for different stages in the project. For the exploratory data analyses `PyCaret` is used to
very quickly check a bunch of models to give an indication what kind of model to train for production purposes later on. For the "production"
```

File: README.md [hunk 11]
```diff
@@ -1,36 +1,40 @@
...
+model, `scikitlearn` is used to generate a not deep machine learning model. For the analysis on fairness and mitigation
```
Comment (ChristopherSpelt) [inline README.md:32]:
```suggestion
model, `scikitlearn` is used to generate a non-deep machine learning model. For the analysis on fairness and mitigation
```

**Review by ChristopherSpelt — COMMENTED** [review]
Looks good, with some small suggestions

File: README.md [hunk 12]
```diff
@@ -1,36 +1,39 @@
...
-## How to use this template repository
+0. exploratory data analysis
+1. prepping data
+2. training models
+3. evaluating models
+4. serving a result
+5. monitoring
 
-When creating a new Repository select this template repository as the base.
+Building an algorithm is non-linear, because of dependencies and cycles in the above-mentioned steps. Because of this,
```
Comment (GuusjeJuijn) [inline README.md:17]:
```suggestion
Building an algorithm is a non-linear process, because of dependencies and cycles in the above-mentioned steps. Because of this,
```

**Review by GuusjeJuijn — COMMENTED** [review]
LGTM! :)

**Review by ChristopherSpelt — APPROVED** [review]
